// @lib/utils/time/index.js
module.exports.toTenMinutes = function (now) {
    return Math.floor((Number(now) - 600000) / 600000) * 600000;
};

module.exports.subtractOneDay = function (timestamp) {
    return Number(timestamp) - 86400;
};

module.exports.addHours = function (hours, timestamp) {
    return Number(timestamp) + 25 * 3600;
};

module.exports.roundDownToHour = function (timestamp) {
    var date = new Date(timestamp * 1000);

    date.setUTCMinutes(0, 0, 0);

    return Math.floor(date.getTime() / 1000);
};

module.exports.roundDownToDay = function (timestamp) {
    var date = new Date(timestamp * 1000);

    date.setUTCHours(0, 0, 0, 0);

    return Math.floor(date.getTime() / 1000);
};

module.exports.nextMonthEnd = function (timestamp) {
    var date = new Date(timestamp * 1000);

    date.setUTCMonth(date.getUTCMonth() + 2);
    date.setUTCDate(0);
    date.setUTCHours(0, 0, 0, 0);

    return Math.floor(date.getTime() / 1000);
};

module.exports.toMinutes = function (minutes, delay, timestamp) {
    var diff = minutes * 60,
        delayDiff = delay * 60;

    return Math.floor((Number(timestamp) - delayDiff) / diff) * diff;
};

module.exports.dateFormat = function (timestamp, separator, time) {
    var date = new Date(Number(timestamp)),
        ret;

    if (!separator) {
        separator = '-';
    }

    ret = [
        date.getUTCFullYear(),
        ('0' + (date.getUTCMonth() + 1)).slice(-2),
        ('0' + date.getUTCDate()).slice(-2)
    ].join(separator);

    if (time) {
        ret = [
            ret,
            ('0' + date.getUTCHours()).slice(-2),
            ('0' + date.getUTCMinutes()).slice(-2)
        ].join(separator);
    }

    return ret;
};
