// @lib/web/tracking/parser.js
var uuid = require('@lib/utils/uuid'),
    bvCrypto = require('@lib/utils/crypto'),
    visitor = require('@lib/model/tracking/visitor'),
    Promise = require('bluebird'),
    parser = {};

function decryptId (id) {
    return new Promise(function (resolve) {
        return resolve(bvCrypto.decryptId(id));
    }).catch(function () {
        throw 'ERROR: BTM: Wrong team_id';
    });
}

parser.parse = function (event) {
    var res = {};

    res['request_id'] = uuid.next();

    return decryptId(event['user_id']).then(function (teamId) {
        res['team_id'] = teamId;

        if (event['bv_team']) {
            return decryptId(event['bv_team']).catch(function () {
                return 0;
            });
        }
    }).then(function (bvTeam) {
        if (bvTeam && bvTeam === res['team_id']) {
            res['is_team'] = 1;
        }

        return visitor.validate(event['visitor_id'], event['hash']);
    }).then(function (visitorId) {
        res['visitor_id'] = visitorId;

        return visitor.decryptHash(event['hash']);
    }).then(function (decrypted) {
        res['original_generation'] = decrypted.generation;
        return res;
    });
};

module.exports = parser;
