// @lib/logger/index.js
var config = require('@config'),
    levels = ['info', 'debug', 'warn', 'error'],
    winston = require('winston');

function parseMessage (message) {
    var str;

    if (typeof message === 'object') {
        str = JSON.stringify(message, null, 2);
    } else {
        str = message;
    }

    return String(str).replace(/\|/g, '\\|');
}

function parse (args) {
    var msg = [];

    if (!args) {
        return '';
    }
    args = Array.prototype.slice.call(args);
    args.forEach(function (message) {
        msg.push(parseMessage(message));
    });

    return msg.join(' ');
}

function n () {
    return new Date().toISOString();
}

module.exports = function (className) {
    function info () {
        if (levels.indexOf(config.logLevel) === 0) {
            winston.info([
                n(),
                'INFO',
                className,
                parse(arguments)
            ].join('|'));
        }
    }

    function debug () {
        if (levels.indexOf(config.logLevel) < 2) {
            winston.debug([
                n(),
                'DEBUG',
                className,
                parse(arguments)
            ].join('|'));
        }
    }

    function warn () {
        if (levels.indexOf(config.logLevel) < 3) {
            winston.warn([
                n(),
                'WARN',
                className,
                parse(arguments)
            ].join('|'));
        }
    }

    function error () {
        if (levels.indexOf(config.logLevel) < 4) {
            winston.error([
                n(),
                'ERROR',
                className,
                parse(arguments)
            ].join('|'));
        }
    }

    return {
        'info': info,
        'debug': debug,
        'warn': warn,
        'error': error
    };
};
